nixos/modules/virtualisation/digital-ocean-init.nix

```
@@ -0,0 +1,63 @@
+{ config, pkgs, lib, ... }:
+with lib;
+{
+  options.virtualisation.digitalOcean.rebuildFromUserData = mkOption {
+    type = bool;
+    enable = true;
+    example = true;
+    description = "Whether to reconfigure the system from Digital Ocean user data";
+  };
+
+  config = {
+    systemd.services.digital-ocean-init = {
+      description = "Reconfigure the system from Digital Ocean uesrdata on startup";
+      wantedBy = [ "multi-user.target" ];
+      unitConfig = {
+        ConditionFileExists = "!/etc/nixos/configuration.nix";
+        After = [ "network-online.target" ];
+        Wants = [ "network-online.target" ];
+      };
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+      };
+      restartIfChanged = false;
+      path = [ pkgs.curl pkgs.gnused pkgs.gnugrep pkgs.systemd config.nix.package config.system.build.nixos-rebuild ];
+      script = ''
+        set -e
+        echo "attempting to fetch configuration from Digital Ocean user data..."
+        export HOME=/root
+        export NIX_PATH=/nix/var/nix/profiles/per-user/root/channels/nixos:nixos-config=/etc/nixos/configuration.nix:/nix/var/nix/profiles/per-user/root/channels
+        userData=$(mktemp)
+        if curl --retry-connrefused --output $userData http://169.254.169.254/metadata/v1/user-data; then
+          # If the user-data looks like it could be a nix expression,
+          # copy it over. Also, look for a magic three-hash comment and set
+          # that as the channel.
+          if sed '/^\(#\|SSH_HOST_.*\)/d' < "$userData" | grep -q '\S'; then
+            channels="$(grep '^###' "$userData" | sed 's|###\s*||')"
+            printf "%s" "$channels" | while read channel; do
+              echo "writing channel: $channel"
+            done
+
+            if [[ -n "$channels" ]]; then
+              printf "%s" "$channels" > /root/.nix-channels
+              nix-channel --update
+            fi
+
+            echo "setting configuration from Digital Ocean user data"
+            cp "$userData" /etc/nixos/configuration.nix
```

```suggestion
            cp "$userData" /etc/nixos/do-userdata.nix
            echo '{ modulesPath, ... }: {
              imports = [
                ./do-userdata.nix
                (modulesPath + "/virtualisation/digital-ocean-config.nix")
              ];
            }' > /etc/nixos/configuration.nix
```
(not necessarily suggesting that the template config be inlined like that but within the constraints of github review ui...)